service/history/workflow/mutable_state_impl.go

```diff
@@ -4374,10 +4374,14 @@ func (ms *MutableStateImpl) UpdateWorkflowStateStatus(
 	return setStateStatus(ms.executionState, state, status)
 }
 
+func (ms *MutableStateImpl) IsDirty() bool {
+	return ms.hBuilder.IsDirty() || len(ms.InsertTasks) > 0
```

Comment:
Shall we check update/delete maps as well? e.g. activity heartbeat case where there's no history event or task.